Extract toAggregateInfo(AggregationNode) helper function

duanmeng opened the pull request:
Refactor HashAggregation to extract toAggregateInfo(AggregationNode). 

Use the new function in StreamingAggregation to reduce copy-paste.

Part of #7665

duanmeng commented:
@mbasmanova Hi Masha, this PR is phase 1 to resolve #7665 could you please help to take a look? Thanks a lot.

After this one is merged I will start phase 2, in which I will add SortedAggregation in the initialization, input, and output processing stages of StreamingAggregation.

mbasmanova reviewed (commented):
@duanmeng Thank you for the refactoring. I wonder if it makes sense to go a step further to refactor Hash/StreamingAggregation to share code for generating AggregateInfos from the plan node.

duanmeng commented:
> @duanmeng Thank you for the refactoring. I wonder if it makes sense to go a step further to refactor Hash/StreamingAggregation to share code for generating AggregateInfos from the plan node.

@mbasmanova Got it, I will refactor this part too.

>Wondering, why don't we populate AggregateInfo.mask?

I used to think that `AggregateInfo.mask` was only used in `SortedAggregations` and populated it in the next PR. I will fix this too.

duanmeng commented:
>Hash/StreamingAggregation to share code for generating AggregateInfos from the plan node.

@mbasmanova Hi Masha, to generate AggregateInfos from the plan we need to,
1. populateInputs,
2. populateFunction,
3. handle and populate sort keys and distinct keys
4. populate masks and intermediate types
5. 
They only share the same code in step 4, and Hash/StreamingAggregation extends `class Operator` directly, we need to insert a middle layer say, `class Aggregation`, and override `populateInputs`, `populateFunction` and `populateSortKeys/distinct` in `Hash/StreamingAggregation`. 
I am not sure it is suitable. What do you think?

duanmeng commented:
@mbasmanova I've put the share code for generating AggregateInfos into `AggregateUtil`, could you please help to take a look? Thanks a lot.

duanmeng commented:
@mbasmanova Hi Masha, I've refactored this PR according to your guidance, could help to take a look? Thanks.

1. Added a helper function `AggregateUtil::toAggregateInfo` to create AggregationInfos from the plan node.
2. Put all the AggregateInfo construction logic in it, and make Hash/StreamingAggregation to use it
3. Fail if AggregationInfo contains unsupported features in StreamingAggregation.

mbasmanova commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
+      std::shared_ptr<core::ExpressionEvaluator>& expressionEvaluator,
+      bool supportLambda = true);
+
+  static std::vector<core::LambdaTypedExprPtr> extractLambdaInputs(
```
Can this be a function in anonymous namespace in the .cpp?

mbasmanova commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
 
+class AggregateUtil {
+ public:
+  static AggregateInfo toAggregateInfo(
```
Please, document this method.

There are a lot of parameter. Would it make changing this method to take AggregationNode and return a list of all AggregateInfos?

mbasmanova commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
+      const RowTypePtr& outputType,
+      core::AggregationNode::Step step,
+      const std::unique_ptr<OperatorCtx>& operatorCtx,
+      memory::MemoryPool* pool,
```
can we remove pool parameter and use operatorCtx->pool()?

mbasmanova commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,21 @@ struct AggregateRegistrationResult {
+      const RowTypePtr& inputType,
+      const RowTypePtr& outputType,
+      core::AggregationNode::Step step,
+      const std::unique_ptr<OperatorCtx>& operatorCtx,
```
maybe pass this as const OperatorCtx&

mbasmanova reviewed (commented):
@duanmeng Looks good % some comments. Please, update PR title and description.

duanmeng commented:
@mbasmanova Hi Masha, I've resolved all the comments and updated the title and description. Could you help to take another look? Thanks.

mbasmanova commented on velox/exec/HashAggregation.cpp:
```diff
@@ -119,64 +66,15 @@ void HashAggregation::initialize() {
+      *aggregationNode_,
+      inputType,
+      outputType_,
+      aggregationNode_->step(),
```
You are already passing aggregationNode_. Why do you need to pass step separately? Same for input and output types.

mbasmanova commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
+  /// @param outputType Output type of the aggregationNode.
+  /// @param step Aggregation step of the aggregationNode.
+  /// @param operatorCtx Operator context.
+  /// @param numKeys Number of group keys.
```
group -> grouping

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
+  /// @param operatorCtx Operator context.
+  /// @param numKeys Number of group keys.
+  /// @param expressionEvaluator Expression evaluation.
+  /// @param isStreaming Indicate this aggregation if streaming or not.
```
Would you explain how this is used?

duanmeng replied:
Added

duanmeng commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
+  /// @param step Aggregation step of the aggregationNode.
+  /// @param operatorCtx Operator context.
+  /// @param numKeys Number of group keys.
+  /// @param expressionEvaluator Expression evaluation.
```
Added

mbasmanova reviewed (commented):
Looks great % small nit.

duanmeng commented on velox/exec/AggregateUtil.h:
```diff
@@ -34,4 +38,29 @@ struct AggregateRegistrationResult {
   }
 };
 
+class AggregateUtil {
```
Sure, done.
